src/test/fuzz/util.h

```diff
@@ -512,4 +512,63 @@ void ReadFromStream(FuzzedDataProvider& fuzzed_data_provider, Stream& stream) no
     }
 }
 
+class FuzzedSocket : public MockableSocket
+{
+    FuzzedDataProvider& m_fuzzed_data_provider;
+
+public:
+    explicit FuzzedSocket(FuzzedDataProvider& fuzzed_data_provider) : m_fuzzed_data_provider{fuzzed_data_provider}
+    {
+    }
+
+    ssize_t send(const char* buf, size_t len, int flags) override
+    {
+        if (m_fuzzed_data_provider.ConsumeBool()) {
+            return len;
+        }
+        return m_fuzzed_data_provider.ConsumeIntegralInRange<ssize_t>(-1, len);
+    }
+
+    ssize_t recv(char* buf, size_t len, int flags) override
+    {
+        if (buf == nullptr || len == 0 || m_fuzzed_data_provider.ConsumeBool()) {
+            return m_fuzzed_data_provider.ConsumeBool() ? 0 : -1;
+        }
+        const std::vector<uint8_t> random_bytes = m_fuzzed_data_provider.ConsumeBytes<uint8_t>(len);
+        if (random_bytes.empty()) {
+            return m_fuzzed_data_provider.ConsumeBool() ? 0 : -1;
+        }
+        std::memcpy(buf, random_bytes.data(), random_bytes.size());
+        if (m_fuzzed_data_provider.ConsumeBool()) {
+            if (len > random_bytes.size()) {
+                std::memset(buf + random_bytes.size(), 0, len - random_bytes.size());
```

Comment:
Why this? `recv(2)` would never set the remaining buffer to `0x0` bytes.

Reply:
This is intentional:

If `m_fuzzed_data_provider.ConsumeBool()` we pretend to read `len` bytes but we only consume `random_bytes.size()` from the fuzzing input and let the remaining bytes be `0x0`.

That way we can satisfy a read of say 4096 bytes even if we only have say 20 bytes of input left.

Makes sense? :)

Reply:
Oh yes, we return `len`. Sorry for the noise.